### PR TITLE
Bug 5428: Missing pkg-config not treated as an error

### PR DESCRIPTION
**Pull request by yadij**
pkg-config is required for building many parts of Squid.
But the PKG_PROG_PKG_CONFIG macro silently ignores
a missing binary. Causing errors far later at build time
which are often mistaken for issues with the library it was
used to detect.

Make pkg-config a mandatory build dependency now
that we rely on PKG_CHECK_MODULES to auto-detect
supported libraries.

### CODE REVIEW

**Comment by yadij** [comment]
Making this as cleared for merge since the issue was already triaged in the bug report and this fix agreed to be added.

**Comment by kinkie** [comment]
whoops, we overlapped. Main difference is my PR (#1902 ) warns, instead of erring.
What is the best way forward?

**Comment by kinkie** [comment]
removed merge flag until we agree on which approach to take

File: configure.ac [hunk 1]
```diff
@@ -117,6 +117,9 @@ dnl Libtool 2.2.6 requires: rm -f
 RM="$RM -f"
 
 PKG_PROG_PKG_CONFIG
+AS_IF([test -z "$PKG_CONFIG"],[
```
Comment (kinkie) [inline configure.ac:120]: the most portable way to test for an empty string is
```
test "x$PKG_CONFIG" = "x"
```

Comment (yadij) [reply configure.ac]: Ah right. done

**Comment by yadij** [comment]
> whoops, we overlapped. Main difference is my PR (#1902 ) warns, instead of erring. What is the best way forward?

IMO error is best for v7 since that version is moving to require it. Older versions the warning may be appropriate if you want to backport a fix.

**Review by rousskov — CHANGES_REQUESTED** [review]
What _required_ Squid features require pkg-config? If they exist, please update PR description to mention one or, better, two of them. Otherwise, we should not require pkg-config.

**Comment by yadij** [comment]
> What _required_ Squid features require pkg-config?

Short answer is:

We all agreed to make `pkg-config` a hard requirement for Squid v7 and I have been removing the code that allows libraries to be detected without it. The addition of this hard error reflects that decision.


Long(er) answer:

.. depends entirely on ones scope / interpretation of the word "required";

 * Developer? or Distributor? or System Admin / User?
 * Does one "require" Linux / BSD firewall integration for NAT / TPROXY?
 * Does one "require" OpenSSL or GnuTLS for HTTPS features?
 * Does one "require" libcap integration for root privileges (hello SELinux)?
 * Does one "require" `make check` to be operational?

... those are the obvious highlighted by `grep PKG_CHECK configure.ac`. The list is changing as more get refactored.

FWIW;  There is already a diagnostic line printed about `pkg-config` being absent (albeit without a "WARNING" panic label). Nobody notices it when the library-specific side-effects grab attention by halting the build. Bug 5428 troubleshoot being just the latest case.

**Review by kinkie — APPROVED** [review]
let's ship it.
Missing pkg-config will result in unexpected build outcomes, better be prescriptive and ensure we have a fully working build environment

**Comment by rousskov** [comment]
> > What _required_ Squid features require pkg-config?
 
> We all agreed to make `pkg-config` a hard requirement for Squid v7

We did not[^1].

[^1]: I hope we agree that ./configure should be testing modules (that normally ship with pkg-config configuration files)  using PKG_CHECK_MODULES(). That design does _not_ make pkg-config program presence a hard requirement for building Squid. PKG_CHECK_MODULES() works correctly in environments without pkg-config program; our `./configure --help` even provides instructions on how to override what pkg-config would do.

> Long(er) answer depends entirely on ones scope / interpretation of the word "required"

In this context, "X is required" means (roughly speaking) "there is no environment where `./configure ... && make` can succeed without X".
 
> ... those are the obvious highlighted by `grep PKG_CHECK configure.ac`.

AFAICT, all modules checked by PKG_CHECK_MODULES() are not "required" (as defined above). The original change request stands.

**Comment by yadij** [comment]
Vetoed by Alex.

**Comment by kinkie** [comment]
@yadij How about just changing AC_MSG_ERROR to AC_MSG_WARN?

**Pull request closed without merging**